starlight/starlight/core/math/matrix4.cpp

```diff
@@ -0,0 +1,117 @@
+#include "matrix4.h"
+#include "utils.h"
+#include <iostream>
+
+Matrix4::Matrix4()
+	: x(Vector4(1.0f, 0.0f, 0.0f, 0.0f)),
+		y(Vector4(0.0f, 1.0f, 0.0f, 0.0f)),
+		z(Vector4(0.0f, 0.0f, 1.0f, 0.0f)),
+		w(Vector4(0.0f, 0.0f, 0.0f, 1.0f))
+{}
+
+Matrix4::Matrix4(const Vector4& v)
+	: x(v.x, 0.0f, 0.0f, 0.0f),
+	y(0.0f, v.y, 0.0f, 0.0f),
+	z(0.0f, 0.0f, v.z, 0.0f),
+	w(0.0f, 0.0f, 0.0f, v.w)
+{}
+
+Matrix4::Matrix4(Vector4 inX, Vector4 inY, Vector4 inZ, Vector4 inW)
+	: x(inX),
+		y(inY),
+		z(inZ),
+		w(inW)
+{}
+
+float Matrix4::Determinant() const
+{
+	// Do I make three temporary matricies so my fingers don't fall off?
+	return x[0] * (y[1] * z[2] - y[2] * z[1])
+		- y[0] * (x[1] * z[2] - x[2] * z[1])
+		+ z[0] * (x[1] * y[2] - x[2] * y[1]);
+}
+
+Vector4 Matrix4::GetScale() const
+{
+	return Vector4(x.x, y.y, z.z, w.w);
+}
+
+Matrix4 Matrix4::Inverse() const
+{
+	float D = Determinant();
+	if (D == 0)
+	{
+		// TODO: Error logging for attempting to invert this
+		return Matrix4();
+	}
+
+	return 1/D * Matrix4(Vector4((y[1] * z[2]) - (y[2] * z[1]),
+						z[1] * x[2] - z[2] * x[1],
+						x[1] * y[2] - x[2] * y[1]),
+				Vector4(z[0] * y[2] - z[2] * y[0],
+						x[0] * z[2] - x[2] * z[0],
+						y[0] * x[2] - y[2] * x[0]),
+				Vector4(y[0] * z[1] - y[1] * z[0],
+						z[0] * x[1] - z[1] * x[0],
+						x[0] * y[1] - x[1] * y[0]));
+}
+
+Matrix4 Matrix4::Transpose() const
+{
+	return Matrix4(Vector4(x[0], y[0], z[0]), Vector4(x[1], y[1], z[1]), Vector4(x[2], y[2], z[2]));
+}
+
+Matrix4 Matrix4::Scale(const Vector4& scalar) const
+{
+	// Cheating with no multiplication since we know we're doing a scale operation
+	return Matrix4(Vector4(x.x * scalar.x, x.y, x.z),
+					Vector4(y.x, y.y * scalar.y, y.z),
+					Vector4(z.x, z.y, z.z * scalar.z));
+}
+
+Matrix4 Matrix4::Rotate(Vector4 axis, float angle) const
```

Comment:
axis should be passed by reference to const since its contents are not being modified